Cleanup boto asg subresources

aneeshusa opened the pull request:
### What does this PR do?
Update boto_asg.absent to autodelete associated CloudWatch Alarms,
if a new kwarg `recurse=True` is passed.
Uses the same `alarms` and `alarms_from_pillar` args as `boto_asg.present`,
to make it easy to switch between the two states.

### What issues does this PR fix or reference?
Requires #43070, #43096.

### Previous Behavior
CloudWatch alarms had to be manually deleted

### New Behavior
CloudWatch alarms can be automatically deleted

### Tests written?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

aneeshusa committed:
Add support for batch Cloudwatch Alarms operations

aneeshusa committed:
Use subkey for boto_asg.present alarms changes

aneeshusa committed:
Delete ASG sub-resources in boto_asg.absent

Add a new kwarg, `recurse`, to expand on the old kwarg `remove_lc`,
which is now deprecated, to enable deletion of sub-resources.

Use the boto_lc.absent state to handle deleting the launch config.
Also, delete any associated CloudWatch alarms.

salt-jenkins commented:
@aneeshusa, thanks for your PR! By analyzing the history of the files in this pull request, we identified @ryan-lane, @lomeroe, @s0undt3ch and @tkwilliams to be potential reviewers.

ryan-lane reviewed (approved):
Reviewed internally. Looks good.

rallytime commented on salt/states/boto_asg.py:
```diff
@@ -851,41 +870,86 @@ def absent(
+            ret['comment'].append('Deleted autoscale group {0}.'.format(name))
+            ret['changes']['asg'] = changes
+
+    if remove_lc or recurse:
```
Can you put a comment in here to inform maintainers what to do when `remove_lc` is to be removed? That will help make the actual removal of the `remove_lc` kwarg much easier in `Neon`. It is unclear to me if the entire block should be removed, or if this line should just be `if recurse:` and the lower `if recurse:` should be collapsed into a single `if` stanza.

It looks like the latter is the way to go, but it's best to be explicit.

rallytime commented on salt/states/boto_cloudwatch_alarm.py:
```diff
@@ -175,15 +175,20 @@ def present(
+        This parameter is ignored if `alarms` is used.
+
+    alarms
+        A list of names of alarms.
```
A version added tag would be good here as well.

rallytime commented on salt/states/boto_asg.py:
```diff
@@ -851,41 +870,86 @@ def absent(
+            profile=profile,
+        )
+        __utils__['state.merge_subreturn'](ret, _ret, subkey='launch_config')
+        if ret['result'] is False:
```
This should be `if _ret['result'] is False:` here, right? I don't see where `ret['result']` could be `False` at this point because it should have already returned.

rallytime commented on salt/states/boto_asg.py:
```diff
@@ -851,41 +870,86 @@ def absent(
+            profile=profile,
+        )
+        __utils__['state.merge_subreturn'](ret, _ret, subkey='alarms')
+        if ret['result'] is False:
```
Same question here as above. I think this check should be looking for `_ret`?

rallytime reviewed (changes requested):
Hi @aneeshusa! This is very nice. Thank you for upstreaming these additions.

I have some small changes (mostly doc related) and a couple of questions. If you could address those, that would be great.

Also, we discussed adding tests for new functions and changes in the boto modules/states in our last call, so I am going to push back on this and request some unit tests be added for these new functions and changes. :)

aneeshusa commented:
Thanks for the review! I'll make the doc updates, add a comment about the `remove_lc` deprecation and add some tests. About checking the result return, the `state.merge_subreturn` util will [update the original return to `False` if the subreturn is `False`](https://github.com/lyft/salt/blob/c9a3126093bd197d1523dc22d5b609193e5710a9/salt/utils/state.py#L184-L186), so we should be able to check either `ret['result'] is False` or `_ret['result'] is False`. I chose to check `ret` somewhat arbitrarily, thought it would be a bit weird to do:
```
if _ret['result']:
    return ret
```
where we check `_ret` but return `ret`. I can switch if you prefer.

rallytime commented:
@aneeshusa Ah, right. Thank you for the clarification there. That is fine I think.

aneeshusa commented:
@rallytime Actually, now that I think about it I think using `_ret['result']` is safer, as sometimes `ret['result']` may not yet be set since the main state is still in progress. I believe for boto_asg.basent `ret['result']` will be set before we call substates, but in general this may not hold. I've updated #43096 to mention this.

rallytime commented:
I was just coming back around here to look at this since the last commit. The following things still need to be fixed up, and then we can get this in:

- [ ] doc updates, as noted in above comments
- [ ] unit tests written
- [ ] waiting on #43096 to be merged, and then this can be rebased.

@aneeshusa Let me know if I am missing anything here on your end. I'm trying to keep track of all of these great changes you've submitted and also keep them moving forward as needed on our end. :)

aneeshusa commented:
I've updated the docs and made a few small fixes, like bumping variables names to at least 3 letters. Since the utils PR is merged, I think the only thing left is for me to finish writing unit tests (and a rebase).

rallytime reviewed (commented):
This looks great. Thank you for your hard work @aneeshusa!

I can review again once those tests are in.

aneeshusa commented:
@rallytime I added unit tests and rebased this to pull in the changes this depends on, ready for review.

rallytime commented:
@aneeshusa Thanks! Looks like there is a minor pylint error and a related test failure:
- https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/14553/violations/
- https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu16-py3/1867/

aneeshusa commented:
Hi @rallytime, I've fixed the pylint issue but I'm not able to reproduce the test failure - it passes when I run it locally. Going to look at this some more but would appreciate it if you have any ideas.

aneeshusa commented:
To follow up, the problem only seems to occur on Ubuntu Xenial with Python3, but I've been unable to reproduce the failure with each of:
- Ubuntu Trusty, Python 2.7
- Ubuntu Trusty, Python 3.4
- Ubuntu Xenial, Python 2.7
- Ubuntu Xenial, Python 3.5
Let me know if you have any ideas about what is going on here.

rallytime commented:
Hey @aneeshusa - sorry for the troubles there. I'll try replicating this, too, and see what I can find.

rallytime commented:
Update: I can't reproduce this when running the test file individually, but I _can_ reproduce it when I run all of the unit tests together. Something running higher up in the unit test list is causing the failure. I'm working on narrowing down where the issue is.

rallytime commented:
Aha! I haven't looked into _why_ yet, but I finally tracked down an easier way to reproduce this test failure.

There is something in the `dockermod` module unit tests, which are executed _before_ the `boto_asg` state unit tests. The failure is reproducible by running the dockermod tests and boto_asg tests sequentially, like this:
```
python3 runtests.py -n unit.modules.test_dockermod -n unit.states.test_boto_asg
```

There is something very strange going on there, because:
1. The error only occurs when the dockermod tests are executed before the boto_asg tests
2. Once the error is triggered, running the `boto_asg` tests individually now fails consistently

So, that's where I would start looking. I won't have time to come back to this likely until later next week, but I wanted to get my findings written down.

@aneeshusa Hopefully that helps tracking this test failure down.

aneeshusa commented:
Thanks for looking, @rallytime. I did some more digging and I think I found the issue. Both `dockermod` and `boto_asg` are loading the `state` utils module.  The `salt-runtests.log` indicates bad magic numbers causing module loading failure when trying to load that module.

When the tests are first run, a `pyc` file is created and dropped into the source location due to the editable install. When the tests are run again, the `pyc` file is found and used for module loading. The `pyc` file staying on disk accounts for the stickiness of the issue.

I find that the issue only occurs if the tests are run under Python 2, and then under Python 3. It suffices to test `boto_asg` both times, `dockermod` is not necessary. It does not seem to occur when going from Python 3 to Python 2, likely due to the introduction of `__pycache__` in Python 3.2.

I dove into the importlib implementation in the cpython source code to look for ways around this. It seems that importlib only relies on file extensions to pick which loader to use - if it sees a `pyc` file but it has the wrong magic numbers, it throws an error instead of skipping it and moving on to the next loader type (e.g. from source). This does not seem to be overridable.

If this is indeed the cause, here are some options for remediation:
- Run the tests entirely without bytecode caching. This is unsatisfactory as it would be nice to be able to use the cache.
- Clear `pyc` files at the beginning of a (full) test run (it might be possible to add a flag to runtests.py for this). This should prevent issues between runs and still allow usage of caching within a run.
- Make LazyLoader smarter with an option to not use the SourcelessFileLoader that loads from pyc files, and enable that option when running tests. Seems too invasive.
- Make LazyLoader much smarter and add bytecode magic number checking, which would require re-implementing importlib functionality. I don't think a brain transplant and upgrade from importlib is a good idea, so I don't recommend this either.

I couldn't find out where the Jenkins configurations are kept, so let me know if I'm off base here and/or how you prefer to work around this.

rallytime commented:
Thanks for looking into this @aneeshusa. That's useful information. I agree that many of those options are quite invasive. I need to think about it more before we consider going down one of those routes.

> I find that the issue only occurs if the tests are run under Python 2, and then under Python 3.

So you are not able to reproduce this issue when you only run tests in Python3?

I ran these tests a couple more times looking through the lens of the PY2 vs PY3 and I think there is something else going on here.  I removed all .pyc files, confirmed that the `__pycache__`s were empty, and re-ran the tests on a clean system only with Python3 and I am still seeing the `bad magic number` errors in the runtests log. Furthermore, the the ubuntu16-py3 tests don't ever run the tests on Python 2 at all, they're all executed with Python 3.

Even when I only run the dockermod test file by itself (after clearing out .pycs of course), those bad magic number errors show themselves in the logs. They don't show themselves when I only run the boto_asg tests (which makes sense).

I think something isn't quite right in the Python 3 support surrounding the dockermod tests, which has gone unnoticed, and now your new tests are exposing the issue.

I think the way to go here is to figure out why those errors are happening and fix it there. I might be able to look into this today. If not, I can come back to this later next week.

rallytime commented:
@aneeshusa I figured it out. I'll submit a fix to your PR.

Commit:
Simplify state util mocking in dockermod unit tests

Loading the utils completely was causing issues with running the
boto_asg state unit tests. We don't need to load all of the state
utils this way for the docker tests, so we should simplify these
calls by using MagicMock instead of the loader.

rallytime commented:
Ok, @aneeshusa I submitted a fix for the test here: https://github.com/lyft/salt/pull/27

But please let me know if that isn't the right place. I can submit something to upstream salt and you can rebase here if desired.

Thanks for upstreaming this and for you patience with the tests.

KumarRajnish commented:
All,

Due to a misconfiguration on our Office365 tenant, all email addressed to your Ruckus Wireless email address (you@ruckuswireless.com) were delivered to a catch all mailbox.  This incident occurred between 11 PM to 8 AM (PST).  IT is delivering these emails with a “Reply-all” rule to all individuals on the email chain.  If you feel you are missing email attachments, please send email to helpdesk@ruckuswireless.com.

We apologize for any inconvenience this may have caused.

Regards,

IT

From: Nicole Thomas [mailto:notifications@github.com]
Sent: Friday, September 15, 2017 7:56 AM
To: saltstack/salt <salt@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [saltstack/salt] Cleanup boto asg subresources (#43327)


Thanks for looking into this @aneeshusa<https://github.com/aneeshusa>. That's useful information. I agree that many of those options are quite invasive. I need to think about it more before we consider going down one of those routes.

I find that the issue only occurs if the tests are run under Python 2, and then under Python 3.

So you are not able to reproduce this issue when you only run tests in Python3?

I ran these tests a couple more times looking through the lens of the PY2 vs PY3 and I think there is something else going on here. I removed all .pyc files, confirmed that the __pycache__s were empty, and re-ran the tests on a clean system only with Python3 and I am still seeing the bad magic number errors in the runtests log. Furthermore, the the ubuntu16-py3 tests don't ever run the tests on Python 2 at all, they're all executed with Python 3.

Even when I only run the dockermod test file by itself (after clearing out .pycs of course), those bad magic number errors show themselves in the logs. They don't show themselves when I only run the boto_asg tests (which makes sense).

I think something isn't quite right in the Python 3 support surrounding the dockermod tests, which has gone unnoticed, and now your new tests are exposing the issue.

I think the way to go here is to figure out why those errors are happening and fix it there. I might be able to look into this today. If not, I can come back to this later next week.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/saltstack/salt/pull/43327#issuecomment-329806492>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AHusM1mq9BaA2KgCM5iJ1mSlMjrMCQ9Dks5sipAIgaJpZM4PKrzR>.

cachedout commented:
Go Go Jenkins!

cachedout commented:
Reading back through this, it appears this is nearly ready to go. There are some minor comments that @rallytime made in an initial review that should be addressed by @aneeshusa  and after that we should be ready to go here.

aneeshusa commented:
Hi @cachedout, I'm back at Lyft now and getting back to these PRs. I think I addressed all of @rallytime's comments - last I remember the last issue was a bad interaction with another test causing the tests to fail, which I'm looking into now. I've rebased this PR on develop to see what Jenkins thinks.

Commit:
Merge branch 'develop' into cleanup-boto_asg-subresources

Commit:
Merge branch 'develop' into cleanup-boto_asg-subresources

rallytime commented:
Hi @aneeshusa - Apologies for not getting back to this sooner. The tests have unfortunately timed out so I am not sure where we ended up on this. 

However, there's now a merge conflict that needs to be resolved. Can you rebase this and fix the conflict when you get a chance?

rallytime commented:
Since we haven't heard back on this for a while, I am going to close this. We can of course re-open this when you're ready to come back to it. This PR will need to be rebased against the `develop` branch and the merge conflicts resolved. Thanks!
